pkg/pillar/cmd/monitor/ipc_server.go

```diff
@@ -0,0 +1,293 @@
+				return
+			}
+			log.Warnf("Error sending response: %v", err)
+		}
+	}
+}
+
+// read request
+func (s *IPCServer) readRequest() (*Request, error) {
+	frame, err := s.codec.ReadFrame()
+	if err != nil {
+		return nil, err
+	}
+	log.Noticef("Received frame: %v", string(frame))
+
+	// following code is used for debugging when #[serde(untagged)] line
+	// is commented out in the rust code in the IpcMessage struct
+	// unmarshal IpcMessage first
+	// var ipcMessage IpcMessage
+	// if err := json.Unmarshal(frame, &ipcMessage); err != nil {
+	// 	return nil, err
+	// }
+
+	var request Request
+	if err := json.Unmarshal(frame, &request); err != nil {
+		return nil, err
+	}
+	return &request, nil
+}
+
+// send response
+func (s *IPCServer) sendResponse(resp *Response) error {
+	return s.SendIpcMessage("Response", resp)
+}
+
+func (s *IPCServer) SendIpcMessage(t string, msg any) error {
+	s.Lock()
+	defer s.Unlock()
+
+	var err error
+
+	if data, err := json.Marshal(msg); err == nil {
+		ipcMessage := IpcMessage{Type: t, Message: json.RawMessage(data)}
+		if data, err = json.Marshal(ipcMessage); err == nil {
+			log.Noticef("Sending IPC message: %s", string(data))
+			_, err = s.codec.Write(data)
+		}
+	}
+	return err
+}
+
+func (r *Request) errResponse(errorText string, err error) *Response {
+	if err != nil {
+		errorText = errorText + ": " + err.Error()
+	}
+	return &Response{
+		Err: errorText,
+		Id:  r.Id,
+	}
+}
+
+func (r *Request) okResponse() *Response {
+	return &Response{
+		Id: r.Id,
+		Ok: "ok",
+	}
+}
+
+func (r *Request) unimplementedResponse() *Response {
+	return r.errResponse("Unimplemented request", nil)
+}
+
+func (r *Request) unknownRequestResponse() *Response {
+	return r.errResponse("Unknown request", nil)
+}
+
+func (r *Request) malformedRequestResponse(err error) *Response {
+	err_message := "Malformed request [" + string(r.RequestData) + "]"
+	return r.errResponse(err_message, err)
+}
+
+func (r *Request) handleRequest(ctx *monitorContext) *Response {
+	switch r.RequestType {
+	case "SetDPC":
+		// Unmarshal the request data
+		var dpc types.DevicePortConfig
+		if err := json.Unmarshal(r.RequestData, &dpc); err == nil {
+			if err := ctx.IPCServer.validateDPC(dpc); err != nil {
+				return r.errResponse("Failed to validate DPC", err)
+			}
+			// unpublish current manual DPC first
+			ctx.pubDevicePortConfig.Unpublish(dpc.Key)
+			// publish the DPC
+			if err := ctx.pubDevicePortConfig.Publish(dpc.Key, dpc); err != nil {
+				return r.errResponse("Failed to publish DPC", err)
+			} else {
+				return r.okResponse()
+			}
+		} else {
+			return r.malformedRequestResponse(err)
+		}
+	case "SetServer":
+		var server string
+		if err := json.Unmarshal(r.RequestData, &server); err == nil {
+			if err := ctx.updateServerFile(server); err != nil {
+				return r.errResponse("Failed to update server file", err)
+			} else {
+				return r.okResponse()
+			}
+		} else {
+			return r.malformedRequestResponse(err)
+		}
```

Comment:
sometimes you check if `err != nil` and then have an else case, sometimes you check if `err == nil` and then have an else case

Reply:
@christoph-zededa why is this bad?

Reply:
I am just wondering what is the reason behind it.

The "bad" thing is that it is deeply nested.

```suggestion
		if err := json.Unmarshal(r.RequestData, &dpc); err != nil {
			return r.malformedRequestResponse(err)
		}
		if err := ctx.IPCServer.validateDPC(dpc); err != nil {
			return r.errResponse("Failed to validate DPC", err)
		}
		// unpublish current manual DPC first
		ctx.pubDevicePortConfig.Unpublish(dpc.Key)
		// publish the DPC
		if err := ctx.pubDevicePortConfig.Publish(dpc.Key, dpc); err != nil {
			return r.errResponse("Failed to publish DPC", err)
		}
		return r.okResponse()

	case "SetServer":
		var server string
		if err := json.Unmarshal(r.RequestData, &server); err != nil {
			return r.malformedRequestResponse(err)
		}
		if err := ctx.updateServerFile(server); err != nil {
			return r.errResponse("Failed to update server file", err)
		}
		return r.okResponse()

```